### ISSUE TITLE
oom caused after use #234's statistics record code

### ISSUE THREAD

**Issue by itherunder** [moka-rs/moka#290]
https://github.com/moka-rs/moka/pull/234

**Comment by tatsuya6502** [comment]
Hi. Thank you for reporting the issue. Can you please provide more information to help us investigate further?

- What is the moka version?
- What is the cache implementation? `sync::Cache`, `sync::SegmentedCache`, or `future::Cache`?
- Is your eviction listener code exactly same to the one in #234?
- What is the `max_capacity` of the cache?
- If you make the `max_capacity` smaller, does the issue still happen?
- What is the average size of the value stored in the cache? (Rough estimate is fine)
- What is the write rate to the cache? (Number of writes per second. Rough number is fine)
    - Are you doing heavy writes? (e.g. millions writes per second, like benchmarking)

**Comment by itherunder** [comment]
> Hi. Thank you for reporting the issue. Can you please provide more information to help us investigate further?
> 
> * What is the moka version?
> * What is the cache implementation? `sync::Cache`, `sync::SegmentedCache`, or `future::Cache`?
> * Is your eviction listener code exactly same to the one in [Statistics for metrics #234](https://github.com/moka-rs/moka/issues/234)?
> * What is the `max_capacity` of the cache?
> * If you make the `max_capacity` smaller, does the issue still happen?
> * What is the average size of the value stored in the cache? (Rough estimate is fine)
> * What is the write rate to the cache? (Number of writes per second. Rough number is fine)
>   
>   * Are you doing heavy writes? (e.g. millions writes per second, like benchmarking)

- 0.11.2
- sync::SegmentedCache / I tried sync::Cache, same oom
- exactly same as #234 
- still happen, I tried to set `max_capacity` just 128
- 20B
- 1000 QPS

**Comment by tatsuya6502** [comment]
@itherunder — Thank you for the info. It was very helpful! I believe I found the root cause of the bug, and the following PR will fix it:

- #295
    - src/sync_base/key_lock.rs#[L33-R33](https://github.com/moka-rs/moka/pull/295/files#diff-e70f778ff17f42108b6e30bd949c34778eb5965957f1a7678be8c7405b221122L33-R33)

When you have chance, please try it to see if the OOM is fixed:

`Cargo.toml`

```toml
[dependencies]
moka = { git = "https://github.com/moka-rs/moka.git", branch = "fix-eviction-listener-key-lock-v011" }
```

I will apply the fix to Moka v0.11.x, v0.10.x and v0.9.x, and hopefully we can release them in this weekend.

**Comment by tatsuya6502** [comment]
> I will apply the fix to Moka v0.11.x, v0.10.x and v0.9.x, and hopefully we can release them in this weekend.

I just published [v0.11.3](https://crates.io/crates/moka/0.11.3), [v0.10.4](https://crates.io/crates/moka/0.10.4) and [v0.9.9](https://crates.io/crates/moka/0.9.9) to crates.io.

Please reopen this issue if the OOM still happens with v0.11.3.
